Add regex string searches and surface geo target list field (Fixes #195, #450)

rfrenchseti opened the pull request:
- Fixes #195
- Fixes #450
- Were any Django, import pipeline, table_schema, or dictionary files modified? Y
  - Database used: opus3_test_200211
  - All Django tests pass: Y
- Were any JavaScript or CSS files modified? Y
  - JSHINT run on all affected files: Y
  - Tested on Chrome, Firefox, Safari, and iOS: Y
    (Remember to test all browser sizes)
  - Tested for race conditions (with delayed API calls if applicable): Y

Description of changes:

Implemented "regex" qtype for strings and added appropriate tests. Made UI do input validation on strings. Implemented "multiple target list" field under Surface Geometry Constraints. Updated documentation.

Known problems:

None.

rfrenchseti committed:
Implement regex string searchs and input validation for strings

rfrenchseti committed:
Add regex to gettingstarted

rfrenchseti committed:
Merge remote-tracking branch 'upstream/master' into 450_regex_searches

rfrenchseti committed:
JSHINT cleanup

rfrenchseti committed:
Add regex tests

rfrenchseti committed:
Making changing string qtypes validate inputs again

rfrenchseti committed:
Make change of string qtype revalidate inputs

rfrenchseti committed:
Merge remote-tracking branch 'origin/450_regex_searches' into 450_regex_searches

rfrenchseti committed:
Implement surface geo target list field

rfrenchseti committed:
Update tests and documentation for surface geo target list field.

rfrenchseti committed:
Merge remote-tracking branch 'upstream/master' into 450_regex_searches

juzen2003 commented on opus/application/apps/search/forms.py:
```diff
@@ -92,7 +92,11 @@ def __init__(self, *args, **kwargs):
+                             'tabindex': 0,
+                             'data-slugname': slug
+                            }),
                     required = False,
```
Do we want to apply the 4 spaces indentation to attrs (https://www.python.org/dev/peps/pep-0008/#indentation)? Because it's an argument passed into widget.

rfrenchseti replied:
Fixed. This whole file is a mess, though, and needs to be cleaned up.

juzen2003 commented on opus/application/apps/search/test_search.py:
```diff
@@ -2830,6 +2830,20 @@ def test__string_query_excludes(self):
         self.assertEqual(params, expected_params)
 
+    def test__string_query_regex(self):
+        "[test_search.py] string_query: string query with qtype contains"
```
with qtype **regex**

rfrenchseti replied:
Fixed

juzen2003 commented on opus/application/static_media/js/hash.js:
```diff
@@ -93,15 +93,15 @@ var o_hash = {
 
                         // If the slug in opus.selections has a valid value (check
-                        // opus.rangeInputFieldsValidation), we push it to the hash.
+                        // opus.InputFieldsValidation), we push it to the hash.
```
Use lower case i (```opus.inputFieldsValidation ```) to stick to JS variable naming convention.

rfrenchseti replied:
Fixed everywhere in JS.

juzen2003 commented on opus/application/static_media/js/opus.js:
```diff
@@ -91,8 +91,8 @@ var opus = {
-    // An object that stores normalize input validation result for each range input.
-    rangeInputFieldsValidation: {},
+    // An object that stores normalizeinput validation result for each input.
+    InputFieldsValidation: {},
```
Use lower case i when declaring ```inputFieldsValidation```.

rfrenchseti replied:
Fixed everywhere in JS.

rfrenchseti committed:
Merge remote-tracking branch 'upstream/master' into 450_regex_searches

rfrenchseti committed:
Fix typo in HTTP404 function

rfrenchseti committed:
Update documentation to say that the central planet is always in the target list.

juzen2003 reviewed (changes requested):
- Issue: 
  - Add an API delay like 3s (OPUS_FAKE_API_DELAYS = 3000)
  - Open up surface geo target selector and multiple target list.
  - In multiple target, select regex in q-type.
  - In surface geo target selector, select "Saturn" and before the search completes:
    - Quickly focus in the input in multiple target, and type "^s.*$" or even just "s" and stay focus in the input with green border.
  - The results and widgets spinner will not go away until we focus out the input (or trigger a change event).
  - UPDATE: the issue can be reproduced with similar steps applying to a combination of mults and input widget, for example:
    - Add an API delay.
    - Open up volume id widget and select regex in qtype.
    - In planet widget, select "Pluto" and before the search completes:
      - Quickly focus in the input of volume id and type something like "\d".
    - The results and widgets spinner will not go away until a change is triggered and done.

- Note: the long multiple target list in autocomplete dropdown will go out of screen.

rfrenchseti committed:
Minor cleanup. Fix bug with normalizeinput reqno.
